src/Features/Core/Portable/PdbSourceDocument/ImplementationAssemblyLookupService.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using System.Xml;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.MetadataAsSource;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+
+namespace Microsoft.CodeAnalysis.PdbSourceDocument
+{
+    [Export(typeof(IImplementationAssemblyLookupService)), Shared]
+    internal class ImplementationAssemblyLookupService : IImplementationAssemblyLookupService
+    {
+        private readonly Dictionary<string, Dictionary<string, string>?> _typeForwardCache = new();
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public ImplementationAssemblyLookupService()
+        {
+        }
+
+        public bool TryFindImplementationAssemblyPath(string referencedDllPath, [NotNullWhen(true)] out string? implementationDllPath)
+        {
+            if (TryNugetLibToRef(referencedDllPath, out implementationDllPath))
+                return true;
+
+            if (TryTargetingPackToSharedSdk(referencedDllPath, out implementationDllPath))
+                return true;
+
+            implementationDllPath = null;
+            return false;
+        }
+
+        public string? FollowTypeForwards(ISymbol symbol, string dllPath, IPdbSourceDocumentLogger? logger)
```

Comment:
>dllPath [](http://example.com/codeflow?start=65&length=7)

Is this a normalized path?

Reply:
All of the methods here expect that they'll be given a full path, and only return normal paths. All of the `..\..\..` craziness is confined to the method that does it, and doesn't escape. Assuming thats what you meant?